### PR TITLE
Use teacher's tokenizer if different from the student's

### PR DESCRIPTION
**Pull request by natuan**
This change fixes the tokenizer when distillation is used and the teacher and student models have different tokenizers. In that case, this fix reuses the teacher's tokenizer with the assumption that it "subsumes" that of the student.

### CODE REVIEW

**Review by bfineran — COMMENTED** [review]
great catch and fix @natuan, left some quick comments.
Larger question for me is - do you forsee cases where the student and teacher have vastly different tokenizers and the "subsume" assumption wouldn't hold?

**Comment by natuan** [comment]
> great catch and fix @natuan, left some quick comments. Larger question for me is - do you forsee cases where the student and teacher have vastly different tokenizers and the "subsume" assumption wouldn't hold?

Probably we might not see this often; more likely when the training objective is different between the two, and the student is smaller (because only then would using a bigger teacher with a different objective make sense). I added a check and throw when the current assumption does not hold.

File: src/sparseml/pytorch/optim/modifier_distillation.py [hunk 1]
```diff
@@ -261,27 +264,30 @@ def loss_update(
...
-            if not self._teacher_input_keys
-            else {key: student_inputs[key] for key in self._teacher_input_keys}
+            (
+                student_inputs
+                if not self._teacher_input_keys
+                else {key: student_inputs[key] for key in self._teacher_input_keys}
+            )
+            if teacher_inputs is None
+            else teacher_inputs
         )
```
[the remark this answers is not in the document]
Comment (natuan) [reply src/sparseml/pytorch/optim/modifier_distillation.py]: Fixed

File: src/sparseml/pytorch/optim/modifier_distillation.py [hunk 2]
```diff
@@ -312,29 +318,13 @@ def loss_update(
                 f"teacher output type of {type(teacher_outputs)}"
             )
 
-        distill_losses = []
```
[the remark this answers is not in the document]
Comment (natuan) [reply src/sparseml/pytorch/optim/modifier_distillation.py:315]: Yes, I didn't change the logic here.

File: src/sparseml/transformers/masked_language_modeling.py [hunk 3]
```diff
@@ -482,6 +498,17 @@ def main():
...
             model_args.tokenizer_name, **tokenizer_kwargs
         )
+    elif model_args.distill_teacher is not None:
+        model_forward_params = list(inspect.signature(model.forward).parameters.keys())
+        teacher_forward_params = list(
+            inspect.signature(teacher.forward).parameters.keys()
+        )
+        diff = [p for p in model_forward_params if p not in teacher_forward_params]
+        if diff:
+            raise RuntimeError("Teacher tokenizer cannot be used for student.")
```
[the remark this answers is not in the document]
Comment (natuan) [reply src/sparseml/transformers/masked_language_modeling.py]: done

**Review by markurtz — APPROVED** [review]

**Review by mgoin — APPROVED** [review]

**Review by bfineran — APPROVED** [review]

**Pull request merged**